### PR TITLE
Convert `InstructionErrors` to coded exceptions

### PR DESCRIPTION
**Pull request by steveluscher**
# Summary

In this PR, we build atop the work in #2213 and introduce coded exceptions for each `InstructionError` returned from the RPC's `sendTransaction` method.

> [!NOTE]
> Because the RPC doesn't return structured errors or error codes, we had to break our own rules in this PR and hardcode a map between the error names and the code numbers. My [first crack](https://gist.github.com/steveluscher/aaa7cbbb5433b1197983908a40860c47#file-fml-ts-L12) at this employed a source code compression scheme that I later deemed too risky for the 250 gzipped bytes it saved. We might consider such a scheme in the future, especially since the next PR will add `InstructionError` to the mix.

# Test Plan

```shell
cd packages/errors
pnpm test:unit:browser
pnpm test:unit:node
```

Addresses #2118.

### CODE REVIEW

**Comment by steveluscher** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2220?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* **#2220** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2220?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#2219** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2219?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2218** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2218?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2217** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2217?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2213** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2213?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2189** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2189?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#2190](https://github.com/solana-labs/solana-web3.js/pull/2190) <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2190?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#2201** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2201?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2200** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2200?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2188** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2188?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2187** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2187?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2186** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2186?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2185** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2184** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2182** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2182?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#2183](https://github.com/solana-labs/solana-web3.js/pull/2183) <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2183?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `master`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2220?utm_source=stack-comment).

File: packages/errors/src/codes.ts [hunk 1]
```diff
@@ -44,6 +44,62 @@ export const SOLANA_ERROR__CODECS_ENCODER_DECODER_SIZE_COMPATIBILITY_MISMATCH =
...
 export const SOLANA_ERROR__CODECS_VARIABLE_SIZE_ENCODER_DECODER_MAX_SIZE_MISMATCH = 37 as const;
 export const SOLANA_ERROR__CODECS_CANNOT_REVERSE_CODEC_OF_VARIABLE_SIZE = 38 as const;
+// Reserve error codes starting with [4615000-4615999] for the Rust enum `InstructionError`
+export const SOLANA_ERROR__INSTRUCTION_ERROR_UNKNOWN = 4615000 as const;
+export const SOLANA_ERROR__INSTRUCTION_ERROR_GENERIC_ERROR = 4615001 as const;
+export const SOLANA_ERROR__INSTRUCTION_ERROR_INVALID_ARGUMENT = 4615002 as const;
+export const SOLANA_ERROR__INSTRUCTION_ERROR_INVALID_INSTRUCTION_DATA = 4615003 as const;
+export const SOLANA_ERROR__INSTRUCTION_ERROR_INVALID_ACCOUNT_DATA = 4615004 as const;
+export const SOLANA_ERROR__INSTRUCTION_ERROR_ACCOUNT_DATA_TOO_SMALL = 4615005 as const;
+export const SOLANA_ERROR__INSTRUCTION_ERROR_INSUFFICIENT_FUNDS = 4615006 as const;
```
Comment (lorisleiva) [inline packages/errors/src/codes.ts:54]: The error code that's gonna save the ecosystem.

File: packages/errors/src/context.ts [hunk 2]
```diff
@@ -38,95 +95,166 @@ export type DefaultUnspecifiedErrorContextToUndefined<T> = {
...
-    };
-    [SOLANA_ERROR__TRANSACTION_ERROR_PROGRAM_EXECUTION_TEMPORARILY_RESTRICTED]: {
-        accountIndex: number;
-    };
-}>;
+export type SolanaErrorContext = DefaultUnspecifiedErrorContextToUndefined<
+    BasicInstructionErrorContext<
+        | typeof SOLANA_ERROR__INSTRUCTION_ERROR_UNKNOWN
+        | typeof SOLANA_ERROR__INSTRUCTION_ERROR_GENERIC_ERROR
+        | typeof SOLANA_ERROR__INSTRUCTION_ERROR_INVALID_ARGUMENT
```
Comment (lorisleiva) [inline packages/errors/src/context.ts:102]: Why is this needed now?

**Review by lorisleiva — APPROVED** [review]
Thank you!

File: packages/errors/src/context.ts (hunk 2 above)
Comment (steveluscher) [reply packages/errors/src/context.ts:102]: I wanted a utility that unconditionally added `index` to _all_ the instruction error contexts.

**Comment by steveluscher** [comment]
### Merge activity

* **Feb 29, 9:52 AM PST**: @steveluscher started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2220).
